### PR TITLE
Feature/execution refactoring poc new object model

### LINKED ISSUE #66: Enhance new object model with additional FlowNode types
#### Describe your issue

There are a number of components, which are not yet covered by the new object model or supported by the object model parser.

These include:
- Text annotations
- Associations
- Intermediate Catch- & Throw- Events of any kind ✅ 
- Subprocesses ✅ 

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by 5ebastianMeier**
Closes #66 

Depends on:
https://github.com/process-engine/process_engine_contracts/pull/33

## What did you change?

This PR takes the handler execution PoC that still used the old data model and adapts it to solely use the new data model.

Also all parsing and handling we cover with integration tests is now fully implemented and works with the integration tests.

Only the persistance is currently not implemented, so process data is only kept in memory.

## How can others test the changes?

Run the integration tests from here: https://github.com/process-engine/process_engine_meta/pull/73

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.)
- [x] I've merged the `develop` branch into my branch before finishing this PR.
- [x] I've **not added any other changes** than the ones described above.
- [x] I've mentioned all **PRs, which relate to this one**

### CODE REVIEW

**Comment by heikomat** [comment]
@5ebastianMeier Please let me know, when this PR is no longer work in progress. I'll gladly take 2 or 3 hours of my time to review this beautiful PR :)

File: README.md [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+
+The ProcessTokenFascade serves to identify the real use cases for a ProcessToken we currently need instead of just passing around the actual token as an object.
+
+Every result that a FlowNode execution produced is saved via the fascade using the method `addResultForFlowNode`.
+
+With a list of results like this, the old token format can be generated by using the method `getOldTokenFormat`. This is merely implemented for compatibility reasons and can later be used to identify all places that still rely on the old ProcessToken concept.
+
+Another compatibility feature is the evaluation of mappers on FlowNodes or on SequenceFlows. These can be evaluated using `evaluateMapperForFlowNode` or `evaluateMapperForSequenceFlow`.
+
+The splitting and merging of ProcessTokens is also done via the fascade using `getProcessTokenFascadeForParallelBranch` and `mergeTokenHistory`.
```
Comment (ElRaptorus) [inline README.md]: Since there are quite a few tasks this facade has to perform, it might make sense to summarize it into a bullet-point list. 

Just a proposal:
```markdown
## ProcessTokenFacade
The `ProcessTokenFacade` identifies and returns any information stored in the `ProcessToken`
that we require for our current use case.

This ensures that each `FlowNode` only gets the information that it actually needs, instead of the
entire process token history.

It performs the following tasks:
* store each FlowNode execution result, using the `addResultForFlowNode` method.
* split process tokens, using `getProcessTokenFascadeForParallelBranch`
* merge process tokens, using `mergeTokenHistory`
* For backwards compatibility:
  *  Get a process token in the old format, using `getOldTokenFormat`
  * evaluate mappers on `FlowNodes` and `SequenceFlows`, using the methods uateMapperForFlowNode` or `evaluateMapperForSequenceFlow`
```

File: README.md [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+* TimerBoundaryEventHandler.execute
+  * ErrorBoundaryEventHandler.execute
+    * ScriptTaskHandler.execute
+
+The order in which the BoundaryEvent-handlers are decorated around the original handler is also important.
+
+In this case the TimerBoundaryEvent-handler has to start its timers as fast as possible.
+If the ScriptTask encounters an error, the ErrorBoundaryEvent-handler has to have the ability to handle it and decide which FlowNode to execute next. On the other hand the ErrorBoundaryEvent-handler would not want to handle an error that related to the TimerBoundaryEvent.
+
+The factory ensures an order like this in a way that prevents conflicts.  However, the original FlowNodeHandler (in this case the ScriptTaskHandler) will always be called last in this chain.
```
[the remark this answers is not in the document]
Comment (5ebastianMeier) [reply README.md]: Your version is simpler and more on point - I wouldn't call it `parent` though, because that might lead to confusion concerning the order of handlers (although it's the parent, it is called last in the call stack)

File: README.md [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+## FlowNodeHandler
+
+The FlowNodeHandler is the base class for all handlers. It provides an abstract `executeIntern`-method for the derived handler to implement its logic.
+
+The base class also offers a private hook that is executed after each FlowNode execution, that is not exposed to derived classes.
```
[the remark this answers is not in the document]
Comment (5ebastianMeier) [reply README.md]: The whole README was not targeted at the end user, but rather at developers extending or working on the process engine. But you're right - an example might make the intention of the hook clearer.

File: README.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+## ProcessTokenFascade
+
+The ProcessTokenFascade serves to identify the real use cases for a ProcessToken we currently need instead of just passing around the actual token as an object.
+
+Every result that a FlowNode execution produced is saved via the fascade using the method `addResultForFlowNode`.
+
+With a list of results like this, the old token format can be generated by using the method `getOldTokenFormat`. This is merely implemented for compatibility reasons and can later be used to identify all places that still rely on the old ProcessToken concept.
```
[the remark this answers is not in the document]
Comment (5ebastianMeier) [reply README.md]: Since this is an internal document and there will surely be several occasions for further refactoring when we fully migrate to the new handler implementation, I'd like to keep the emphasis for now.

It explains an important part of how somebody can understand where the token is used and how.

The README was merely meant as a crash course for the new implementation and the current state of the implementation.

File: README.md [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+
+Another compatibility feature is the evaluation of mappers on FlowNodes or on SequenceFlows. These can be evaluated using `evaluateMapperForFlowNode` or `evaluateMapperForSequenceFlow`.
+
+The splitting and merging of ProcessTokens is also done via the fascade using `getProcessTokenFascadeForParallelBranch` and `mergeTokenHistory`.
+
+## ProcessModelFascade
+
+Whenever an element of the process model is needed in a specific context, it is queried for using the ProcessModelFascade.
```
[the remark this answers is not in the document]
Comment (5ebastianMeier) [reply README.md]: The word `context` is a bit misleading here - I meant the use case (e.g. give my the next flow node after the given flow node)

File: src/new_model/runtime/engine/handler/call_activity_handler.ts [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+    super();
+
+    this._consumerApiService = consumerApiService;
+  }
+
+  private get consumerApiService(): IConsumerApiService {
+    return this._consumerApiService;
+  }
+
+  protected async executeIntern(callActivityNode: Model.Activities.CallActivity,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/call_activity_handler.ts]: *`executeInternally`

An "Intern" is (usually) not something you'd wish to execute: https://www.dict.cc/englisch-deutsch/intern.html Although I guess there are always some exceptions. 😁

File: src/new_model/runtime/engine/handler/end_event_handler.ts [hunk 7]
```diff
@@ -0,0 +1,13 @@
+import { IExecutionContextFacade, IFlowNodeHandlerFactory, IProcessModelFacade,
+  IProcessTokenFacade, Model, NextFlowNodeInfo, Runtime} from '@process-engine/process_engine_contracts';
+import { FlowNodeHandler } from './index';
+
+export class EndEventHandler extends FlowNodeHandler<Model.Events.EndEvent> {
+
+  protected async executeIntern(flowNode: Model.Events.EndEvent,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/end_event_handler.ts]: *`executeInternally`

Seriously, this one of the most epic typos ever 😆

File: src/new_model/runtime/engine/handler/error_boundary_event_handler.ts [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+  constructor(decoratedHandler: FlowNodeHandler<Model.Base.FlowNode>) {
+    super();
+    this._decoratedHandler = decoratedHandler;
+  }
+
+  private get decoratedHandler(): FlowNodeHandler<Model.Base.FlowNode> {
+    return this._decoratedHandler;
+  }
+
+  protected async executeIntern(flowNode: Model.Events.BoundaryEvent,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/error_boundary_event_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/exclusive_gateway_handler.ts [hunk 9]
```diff
@@ -0,0 +1,64 @@
+import { IExecutionContextFacade, IProcessModelFacade, IProcessTokenFacade, Model,
+  NextFlowNodeInfo, Runtime } from '@process-engine/process_engine_contracts';
+import { FlowNodeHandler } from './flow_node_handler';
+
+export class ExclusiveGatewayHandler extends FlowNodeHandler<Model.Gateways.ExclusiveGateway> {
+
+  protected async executeIntern(flowNode: Model.Gateways.ExclusiveGateway,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/exclusive_gateway_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/flow_node_handler.ts [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+                       processTokenFacade: IProcessTokenFacade,
+                       processModelFacade: IProcessModelFacade,
+                       executionContextFacade: IExecutionContextFacade): Promise<NextFlowNodeInfo> {
+
+    let nextFlowNode: NextFlowNodeInfo;
+
+    try {
+
+      // executeIntern is the method where derived handlers can implement their logic
+      nextFlowNode = await this.executeIntern(flowNode, processTokenFacade, processModelFacade, executionContextFacade);
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/flow_node_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/flow_node_handler.ts [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+    } catch (error) {
+      // TODO: (SM) this is only to support the old implementation
+      //            I would like to set no token result or further specify it to be an error to avoid confusion
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/flow_node_handler.ts]: Agreed. 👍

File: src/new_model/runtime/engine/handler/flow_node_handler.ts [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+    if (!nextFlowNode) {
+      throw new Error(`Next flow node after node with id "${flowNode.id}" could not be found.`);
+    }
+
+    await this.afterExecute(flowNode, nextFlowNode.flowNode, processTokenFacade, processModelFacade);
+
+    return nextFlowNode;
+  }
+
+  protected async abstract executeIntern(flowNode: TFlowNode,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/flow_node_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/flow_node_handler_factory.ts [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+  constructor(container: IContainer) {
+    this._container = container;
+  }
+
+  private get container(): IContainer {
+    return this._container;
+  }
+
+  public async create<TFlowNode extends Model.Base.FlowNode>(flowNode: TFlowNode,
+                                                             processModelFascade: IProcessModelFacade): Promise<IFlowNodeHandler<TFlowNode>> {
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/flow_node_handler_factory.ts]: *Facade

File: src/new_model/runtime/engine/handler/flow_node_handler_factory.ts [hunk 14]
```diff
@@ -0,0 +1,146 @@
...
+
+  private _getEventDefinitionType(boundaryEventNode: Model.Events.BoundaryEvent): BoundaryEventDefinitionType {
+    if (boundaryEventNode.errorEventDefinition) {
+      return BoundaryEventDefinitionType.Error;
+    } else if (boundaryEventNode.messageEventDefinition) {
+      return BoundaryEventDefinitionType.Message;
+    } else if (boundaryEventNode.signalEventDefinition) {
+      return BoundaryEventDefinitionType.Signal;
+    } else if (boundaryEventNode.timerEventDefinition) {
+      return BoundaryEventDefinitionType.Timer;
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/flow_node_handler_factory.ts]: Not really fond of this pattern.

```TypeScript
if (conditionA) {
  return resultA;
} else if (conditionB) {
  return resultB;
}
```

Better use 

```TypeScript
if (conditionA) {
  return resultA;
}

if (conditionB) {
  return resultB;
}
```

or: 

```TypeScript

let finalResult;

if (conditionA) {
  result = resultA;
} else if (conditionB) {
  result = resultB;
}

return finalResult;
```

Didn't we have a corresponding rule for this in our Styleguides anyway?

I distinctly remember that at least ESLint used to complain about this.

File: src/new_model/runtime/engine/handler/flow_node_handler_factory.ts [hunk 15]
```diff
@@ -0,0 +1,146 @@
...
+    // not errors of other boundary events or so
+
+    boundaryEvents.sort((eventA: Model.Events.BoundaryEvent, eventB: Model.Events.BoundaryEvent) => {
+
+      const eventAType: number = this._getEventDefinitionType(eventA);
+      const eventBType: number = this._getEventDefinitionType(eventB);
+
+      if (eventAType < eventBType) {
+        return -1;
+      } else if (eventAType > eventBType) {
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/flow_node_handler_factory.ts]: Like above, better use: 

```TypeScript
if (eventAType < eventBType) {
  return -1;
}

if (eventAType > eventBType) {
  return 1;
}

return 0;
```

File: src/new_model/runtime/engine/handler/intermediate_catch_event_handler.ts [hunk 16]
```diff
@@ -0,0 +1,15 @@
+import { IExecutionContextFacade, IProcessModelFacade, IProcessTokenFacade,
+  Model, NextFlowNodeInfo, Runtime } from '@process-engine/process_engine_contracts';
+import { FlowNodeHandler } from './index';
+
+export class IntermediateCatchEventHandler extends FlowNodeHandler<Model.Events.Event> {
+  protected async executeIntern(flowNode: Model.Events.Event,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/intermediate_catch_event_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/intermediate_throw_event_handler.ts [hunk 17]
```diff
@@ -0,0 +1,15 @@
+import { IExecutionContextFacade, IFlowNodeHandlerFactory,
+  IProcessModelFacade, IProcessTokenFacade, Model, NextFlowNodeInfo, Runtime } from '@process-engine/process_engine_contracts';
+import { FlowNodeHandler } from './index';
+
+export class IntermediateThrowEventHandler extends FlowNodeHandler<Model.Events.Event> {
+  protected async executeIntern(flowNode: Model.Events.Event,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/intermediate_throw_event_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/flow_node_handler_factory.ts [hunk 18]
```diff
@@ -0,0 +1,146 @@
...
+        throw Error(`Es konnte kein FlowNodeHandler für den FlowNodeType ${type} gefunden werden.`);
+    }
+  }
+
+  private async _createHandler<TFlowNode extends Model.Base.FlowNode>(registrationKey: string): Promise<IFlowNodeHandler<TFlowNode>> {
+    return this.container.resolveAsync<IFlowNodeHandler<TFlowNode>>(registrationKey);
+  }
+
+  private async _decorateWithBoundaryEventHandlers<TFlowNode extends Model.Base.FlowNode>(boundaryEvents: Array<Model.Events.BoundaryEvent>,
+                                                                                          processModelFascade: IProcessModelFacade,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/flow_node_handler_factory.ts]: *Facade

Is that parameter even required here? It's not used anywhere.

File: src/new_model/runtime/engine/handler/parallel_gateway_handler.ts [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+  constructor(flowNodeHandlerFactory: IFlowNodeHandlerFactory) {
+    super();
+    this._flowNodeHandlerFactory = flowNodeHandlerFactory;
+  }
+
+  private get flowNodeHandlerFactory(): IFlowNodeHandlerFactory {
+    return this._flowNodeHandlerFactory;
+  }
+
+  protected async executeIntern(flowNode: Model.Gateways.ParallelGateway,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/parallel_gateway_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/script_task_handler.ts [hunk 20]
```diff
@@ -0,0 +1,41 @@
+import { ExecutionContext, IToPojoOptions } from '@essential-projects/core_contracts';
+import { IExecutionContextFacade, IFlowNodeHandlerFactory, IProcessModelFacade, IProcessTokenFacade,
+  Model, NextFlowNodeInfo, Runtime } from '@process-engine/process_engine_contracts';
+import { FlowNodeHandler } from './index';
+
+export class ScriptTaskHandler extends FlowNodeHandler<Model.Activities.ScriptTask> {
+
+  protected async executeIntern(scriptTask: Model.Activities.ScriptTask,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/script_task_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/script_task_handler.ts [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+                                processTokenFacade: IProcessTokenFacade,
+                                processModelFacade: IProcessModelFacade,
+                                executionContextFacade: IExecutionContextFacade): Promise<NextFlowNodeInfo> {
+
+    const script: string = scriptTask.script;
+    const context: ExecutionContext = undefined; // TODO: context needed
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/script_task_handler.ts]: Why not use the `executionContextFacade` to get this?

File: src/new_model/runtime/engine/handler/service_task_handler.ts [hunk 22]
```diff
@@ -0,0 +1,65 @@
...
+
+  private get container(): IContainer {
+    return this._container;
+  }
+
+  private get invoker(): IInvoker {
+    return this._invoker;
+  }
+
+  protected async executeIntern(serviceTaskNode: Model.Activities.ServiceTask,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/service_task_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/service_task_handler.ts [hunk 23]
```diff
@@ -0,0 +1,65 @@
...
+                                processTokenFacade: IProcessTokenFacade,
+                                processModelFacade: IProcessModelFacade,
+                                executionContextFacade: IExecutionContextFacade): Promise<NextFlowNodeInfo> {
+
+    const context: ExecutionContext = executionContextFacade.getExecutionContext();
+    const isMethodInvocation: boolean = serviceTaskNode.invocation instanceof Model.Activities.MethodInvocation;
+    const tokenData: any = await processTokenFacade.getOldTokenFormat();
+
+    if (isMethodInvocation) {
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/service_task_handler.ts]: I  think the ScriptTask component really demonstrates the pure awesomeness of this new approach.  ❤️

File: src/new_model/runtime/engine/handler/service_task_handler.ts [hunk 24]
```diff
@@ -0,0 +1,65 @@
...
+
+      const invocation: Model.Activities.MethodInvocation = serviceTaskNode.invocation as Model.Activities.MethodInvocation;
+
+      const serviceInstance: any = await this.container.resolveAsync(invocation.module);
+
+      // tslint:disable-next-line:max-line-length
+      const argumentsToPassThrough: Array<any> = (new Function('context', 'token', `return ${invocation.params}`)).call(tokenData, context, tokenData) || [];
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/service_task_handler.ts]: Maybe split into multiple statements and/or lines, instead of disabling the rule? 😉

File: src/new_model/runtime/engine/handler/start_event_handler.ts [hunk 25]
```diff
@@ -0,0 +1,15 @@
+import { IExecutionContextFacade, IProcessModelFacade, IProcessTokenFacade, Model, NextFlowNodeInfo,
+  Runtime } from '@process-engine/process_engine_contracts';
+import { FlowNodeHandler } from './index';
+
+export class StartEventHandler extends FlowNodeHandler<Model.Events.StartEvent> {
+
+  protected async executeIntern(flowNode: Model.Events.StartEvent,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/start_event_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/sub_process_handler.ts [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+  constructor(flowNodeHandlerFactory: IFlowNodeHandlerFactory) {
+    super();
+    this._flowNodeHandlerFactory = flowNodeHandlerFactory;
+  }
+
+  private get flowNodeHandlerFactory(): IFlowNodeHandlerFactory {
+    return this._flowNodeHandlerFactory;
+  }
+
+  protected async executeIntern(subProcessNode: Model.Activities.SubProcess,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/sub_process_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/timer_boundary_event_handler.ts [hunk 27]
```diff
@@ -0,0 +1,216 @@
+// tslint:disable:max-line-length
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/timer_boundary_event_handler.ts]: Don't.

File: src/new_model/runtime/engine/handler/timer_boundary_event_handler.ts [hunk 28]
```diff
@@ -0,0 +1,216 @@
...
+
+  private get iamService(): IIamService {
+    return this._iamService;
+  }
+
+  private get decoratedHandler(): FlowNodeHandler<Model.Base.FlowNode> {
+    return this._decoratedHandler;
+  }
+
+  protected async executeIntern(flowNode: Model.Base.FlowNode,
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/timer_boundary_event_handler.ts]: *`executeInternally`

File: src/new_model/runtime/engine/handler/timer_boundary_event_handler.ts [hunk 29]
```diff
@@ -0,0 +1,216 @@
...
+  }
+
+  private _parseTimerDefinitionType(eventDefinition: any): TimerDefinitionType {
+    if (eventDefinition['bpmn:timeDuration']) {
+      return TimerDefinitionType.duration;
+    }
+    if (eventDefinition['bpmn:timeCycle']) {
+      return TimerDefinitionType.cycle;
+    }
+    if (eventDefinition['bpmn:timeDate']) {
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/timer_boundary_event_handler.ts]: Not sure, but aren't these tags stored in the contracts package?

If so, I'd prefer to use them instead of the magic strings.

File: src/new_model/runtime/engine/handler/timer_boundary_event_handler.ts [hunk 30]
```diff
@@ -0,0 +1,216 @@
...
+      return TimerDefinitionType.date;
+    }
+
+    return undefined;
+  }
+
+  private _parseTimerDefinitionValue(eventDefinition: any): string {
+    if (eventDefinition['bpmn:timeDuration']) {
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/handler/timer_boundary_event_handler.ts]: See above, please check if these tags are stored in the contracts package.

File: src/new_model/runtime/engine/process_model_facade.ts [hunk 31]
```diff
@@ -0,0 +1,128 @@
...
+    });
+  }
+
+  public getOutgoingSequenceFlowsFor(flowNodeId: string): Array<Model.Types.SequenceFlow> {
+    return this.processDefinition.sequenceFlows.filter((sequenceFlow: Model.Types.SequenceFlow) => {
+      return sequenceFlow.sourceRef === flowNodeId;
+    });
+  }
+
+  // TODO: support of new Split Gateway in Branch
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/process_model_facade.ts]: Please clarify, what is missing here to support this?

File: src/new_model/runtime/engine/sub_process_model_fascade.ts [hunk 32]
```diff
@@ -0,0 +1,47 @@
+import { IProcessModelFacade, Model } from '@process-engine/process_engine_contracts';
+import { ProcessModelFacade } from './process_model_facade';
```
Comment (ElRaptorus) [inline src/new_model/runtime/engine/sub_process_model_fascade.ts]: Duplicated file. 

Please remove this one, as its name contains the `fascade` typo.

File: src/process_engine_service.ts [hunk 33]
```diff
@@ -522,27 +535,13 @@ export class ProcessEngineService implements IProcessEngineService {
       throw new Error(`Couldn't execute process: neither id nor key of processDefinition is provided`);
     }
 
-    let processDefinition: IProcessDefEntity;
-    if (id !== undefined) {
-      processDefinition = await this.processDefEntityTypeService.getProcessDefinitionById(context, id, version);
-    } else {
-      processDefinition = await this.processDefEntityTypeService.getProcessDefinitionByKey(context, key, version);
-    }
+    const process: Model.Types.Process = await this.processEngineStorageService.getProcess(key);
```
Comment (ElRaptorus) [inline src/process_engine_service.ts]: Am I correct in assuming that retrieving a process model by its ID is intentionally no longer supported?

File: src/process_engine_service.ts [hunk 34]
```diff
@@ -522,27 +535,13 @@ export class ProcessEngineService implements IProcessEngineService {
...
 
-    if (!processDefinition) {
-      throw new Error(`couldn't execute process: no processDefinition with key ${key} or id ${id} was found`);
+    if (!process) {
+      throw new Error(`couldn't execute process: no process with id "${id}" was found`);
```
Comment (ElRaptorus) [inline src/process_engine_service.ts]: I think this should say: `... no process with key "${key}" was found`, since you only retrieve the process model by its key, not by its ID.

File: src/process_engine_service.ts [hunk 35]
```diff
@@ -599,52 +598,55 @@ export class ProcessEngineService implements IProcessEngineService {
...
+      // });
 
-      await this.invoker.invoke(processInstance, 'start', undefined, context, context, {
-        initialToken: initialToken,
-      });
-    });
+      // await this.invoker.invoke(processInstance, 'start', undefined, context, context, {
+      //   initialToken: initialToken,
+      // });
+      const tokenResult: any = await this._executeProcessService.start(context, process, initialToken);
```
Comment (ElRaptorus) [inline src/process_engine_service.ts]: I can't believe how awesomely succinct this is. 😃 

But please remove all the commented-out code, if it is no longer necessary.

File: src/new_model/model/parser/event_parser.ts [hunk 36]
```diff
@@ -16,7 +16,55 @@ export function parseEventsFromProcessData(processData: any): Array<Model.Events
...
+
+}
+
+function assignEventDefinition(event: any, eventRaw: any, eventRawTagName: string, targetPropertyName: string): void {
+  let eventDefinitonValue: any = eventRaw[eventRawTagName];
+  if (eventDefinitonValue === '') {
+    eventDefinitonValue = {};
+  }
+  if (eventDefinitonValue !== undefined && eventDefinitonValue !== null) {
+    event[targetPropertyName] = eventDefinitonValue;
```
[the remark this answers is not in the document]
Comment (5ebastianMeier) [reply src/new_model/model/parser/event_parser.ts]: I checked this again and I think the references are currently never used.

The references seem to be necessary when an event definition is not directly contained in the node referencing it, e.g. when a `BoundaryEvent` doesn't contain an `ErrorEventDefinition`.

The BPMN spec describes the use case for the references for when the event definitions are declared on a global scope. That means the event definition could be declared on the whole BPMN file (referencable by each contained process) or event in another external file.

As we currently don't use references like that, I removed the references completely. That means the parsed event definition doesn't contain a reference, but the actual event definition data.

File: src/new_model/model/parser/process_parser.ts [hunk 37]
```diff
@@ -0,0 +1,79 @@
...
+  }
+
+  return processes;
+}
+
+// TODO: The following elements are not supported yet:
+// - Text annotations
+// - Associations
+// - Intermediate Catch- & Throw- Events of any kind
+// - Subprocess
```
[the remark this answers is not in the document]
Comment (5ebastianMeier) [reply src/new_model/model/parser/process_parser.ts]: Definitely. There might still be edge cases concerning the mentioned elements, that are not being implemented yet.

But the part relevant to our current integration test certainly works, so revisiting this later seems fine.

File: src/new_model/runtime/engine/process_model_facade.ts (hunk 31 above)
Comment (5ebastianMeier) [reply src/new_model/runtime/engine/process_model_facade.ts]: This is actually a TODO-comment I copied from @roblen45.

Afaik the current implementation takes the first `ParallelGateway` that looks like a join gateway when trying to evaluate how far a branch gets executed.

I've updated the TODO-comment to further specify this.

File: src/process_engine_service.ts (hunk 33 above)
Comment (5ebastianMeier) [reply src/process_engine_service.ts]: Yes.

File: src/process_engine_service.ts (hunk 34 above)
Comment (5ebastianMeier) [reply src/process_engine_service.ts]: Based on our old implementation and the old vocabulary you're right. According to the BPMN spec there is no such thing as a key, it is just called the ID. That's why I decided to break with the old style and use the wording `ID` here.

File: src/new_model/runtime/engine/handler/timer_boundary_event_handler.ts (hunk 29 above)
Comment (5ebastianMeier) [reply src/new_model/runtime/engine/handler/timer_boundary_event_handler.ts]: Currently these are not contained in neither `process_engine_contracts`, nor `core_contracts`.

Even in the new handler PoC, these are only ever used in the `TimerBoundaryEventHandler`.

I'm with you that its a good practice to place this in its own enum, but we could discuss to not export the enum for external use until we have a use case for that.

Comment (ElRaptorus) [reply src/new_model/runtime/engine/handler/timer_boundary_event_handler.ts]: Sounds good. Thx for checking. :+1:
